Build Tooling

jlhughes opened the pull request.

jlhughes committed:
Merge branch 'development' of github.com:kgrgreer/foam3 into BuildSuppressTest

jlhughes committed:
Merge branch 'development' of github.com:kgrgreer/foam3 into BuildSuppressTest

jlhughes committed:
tools - break build.js into Tooling poms

jlhughes committed:
accidentally committed

jlhughes committed:
tools - cleanup, ensure test clean is not run when not in test

jlhughes committed:
Merge branch 'development' of github.com:kgrgreer/foam3 into BuildTooling

jlhughes committed:
tooling - update duplicate warning text

jlhughes committed:
tooling - add missing 'this' for build provided function

jlhughes committed:
update copy

jlhughes committed:
build doc - rework text

jlhughes committed:
doc - update

jlhughes committed:
doc - add TODOs

jlhughes committed:
build - control execSync stdio with VERBOSE - either inherit or ignore

jlhughes committed:
Merge branch 'development' of github.com:kgrgreer/foam3 into BuildTooling

kgrgreer commented:
LGTM

adamvy commented on tools/ExampleTooling.js:
```diff
@@ -0,0 +1,35 @@
+const pmake    = require('./pmake.js');
+const { comma, copyDir, copyFile, emptyDir, ensureDir, exec, execSync, exportEnvs, flag, info, rmdir, rmfile, warning } = require('./buildlib');
+
+foam.POM({
```
might be nice to move the example to an examples subdir

adamvy commented on tools/JSTooling.js:
```diff
@@ -0,0 +1,51 @@
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+const pmake    = require('./pmake.js');
```
as discussed in chat, we should find a way to inject these stuff into the tasks directly instead of needing to use require().

That way these pom files can look the same as the pom files in the source tree and still be loaded on in other environmetns that don't provide require() (such as web)


I think the existing build.js tried to achieve this by setting "this" to an object that contains all the buildlib functions when running the pom tasks so that the tasks can do this `this.execSync` and whatnot

adamvy commented on tools/Build.flow:
```diff
@@ -0,0 +1,59 @@
+
+<p><b>POMs</b> are JSON files which identify source code, application configuration, and build instructions. They are hierarchical and can exist at any level of a code repository.</p>
+
+<p>There are two types of POMs:
```
Let's expand or rephrase this.

Conceptually there are two types of fields to the pom, tooling fields which set up tasks and environment variables for the build.js process, and build fields which declare build inputs such as files/javaFiles etc.

It would also be useful to add a section about how environment variables work.

From what I gather, environment variables are delcared with default values but can be overriden by environment variables that exist when you run the code?  And environment variables are passed down into tools run by the build process somehow?

adamvy reviewed (approved):
One further thought, I wonder if this is a good opporunity to break backwards compatability with build.sh and rework the arguments.

Right now arguments can only be one letter and we are running out.  It might be good to support classic GNU getopt style arguments where they can have a long version and a short version

```
foam.POM({
  args: [ 'build-jar', 'j', 'Build a JAR...', ... ]
})
```

can be invoked as

```
$ node tools/build.js -f
# Or
$ node tools/build.js --build-jar
```

jlhughes committed:
cpature ideas, comments, critique for future work

jlhughes committed:
Merge branch 'development' of github.com:kgrgreer/foam3 into BuildTooling

jlhughes committed:
tools - export all functions required by Tooling poms and then bind the build.js 'this' when calling pmake so the Tooling poms do not require require(), but instead can direclty access buildlib functions via 'this'

jlhughes committed:
example - move ExampleTooling to example directory and update flow doc now that require() is resolved

jlhughes committed:
doc - removed completed items

jlhughes committed:
tool - partial support for GNU work style arguments - presently restricted to task names

jlhughes committed:
css - move css from np controller to this Document - a more appropriate home

jlhughes committed:
tools - refactored to remove need for require() in poms. Support gnu getopt style options.

jlhughes committed:
Merge branch 'development' of github.com:kgrgreer/foam3 into BuildTooling

jlhughes committed:
tools - improve usage and topic output

jlhughes committed:
tools - Refactor options and tasks from arrays into objects with named parameters. With named parameters, array indexing is replaced with the named parameter, which is easier to read, maintain, and is less fragile should the option or task objects be modified

jlhughes committed:
Merge branch 'development' of github.com:kgrgreer/foam3 into BuildTooling

jlhughes committed:
tools - ensure all environment variables are displayed by showReport

jlhughes committed:
Merge branch 'development' of github.com:kgrgreer/foam3 into BuildTooling

jlhughes committed:
Merge branch 'development' of github.com:kgrgreer/foam3 into BuildTooling

jlhughes committed:
tools - do not overwrite variables set by command line options with pom envs

jlhughes committed:
tools - more cleanup, consolidation, and doc

jlhughes committed:
tools - reduce number of tooling environment variables

jlhughes committed:
tools - revert move of FOAM_BIN_VERSION

jlhughes committed:
tools - remove duplication of ENV and ENVS

jlhughes committed:
Merge branch 'development' of github.com:kgrgreer/foam3 into BuildTooling

jlhughes committed:
tools - finalized on flag toggling logic. javaTests no longer require any environment variables

jlhughes committed:
Merge branch 'development' of github.com:kgrgreer/foam3 into BuildTooling

jlhughes committed:
Boot - build support for main 'args'

Use build option javaMainClass to set indicate the application entry point.
Use build option javaMainArgs to pass arguments to the main class
Replaced old references to Nanos with CORE (missed in migration)

jlhughes committed:
build - move bootscript to Java main arg.  Fix javaTest setup - out of order clean

jlhughes committed:
Merge branch 'development' of github.com:kgrgreer/foam3 into BuildTooling

jlhughes committed:
build - split building java manifest into tasks each tooling pom can contribute to.

jlhughes committed:
JavaTooling - hack to set test bootscript - task dependencies are being clobbered, when defined in multiple tooling poms.

jlhughes committed:
tooling - refactor TASKS to create stack of same named tooling tasks and call each depedency list - previously only dependency list of first task was used

jlhughes committed:
build - pomTasks - execute in reverse pom order giving poms higher in hierachy abilty to modify/replace values set earlier.

jlhughes committed:
tools - add os.platform option. Add Project tooling for creating a new project. Support specifying tooling poms in text file tools/defaultTooling

jlhughes committed:
Merge branch 'development' of github.com:kgrgreer/foam3 into BuildTooling

jlhughes committed:
build - correct call to buildlibs.bool, add PLATFORM (os.platform) option, and add tooling Project for creating new foam based projects

jlhughes committed:
build - fix flow doc invalid tag closing

jlhughes committed:
build - revert development work on VERBOSE - it was incomplete and failing

jlhughes committed:
build - correct processing of single char arguments when both upper and lower case present

jlhughes committed:
build - chage error to warning for missing root pom, this allows ProjectTooling to create the root pom

jlhughes committed:
build - update flow doc regarding ProjectTooling

jlhughes committed:
build - Add tooling to create a complete demo app

jlhughes committed:
build - rework Project to support explicit appName and modelName

jlhughes committed:
Merge branch 'development' of github.com:kgrgreer/foam3 into BuildTooling

jlhughes committed:
build - more usage examples. when searching for deployment journal, if not found in application, test foam3/deployment
